src/bio.c

```diff
@@ -247,17 +259,29 @@ void *bioProcessBackgroundJobs(void *arg) {
 
         /* Lock again before reiterating the loop, if there are no longer
          * jobs to process we'll block again in pthread_cond_wait(). */
-        pthread_mutex_lock(&bio_mutex[type]);
-        listDelNode(bio_jobs[type],ln);
+        pthread_mutex_lock(&bio_mutex[worker]);
+        listDelNode(bio_jobs[worker], ln);
     }
 }
 
 /* Return the number of pending jobs of the specified type. */
 unsigned long bioPendingJobsOfType(int type) {
-    unsigned long long val;
-    pthread_mutex_lock(&bio_mutex[type]);
-    val = listLength(bio_jobs[type]);
-    pthread_mutex_unlock(&bio_mutex[type]);
+    unsigned long long val = 0;
+    unsigned long worker = bio_job_to_worker[type];
+    listIter li;
+    listNode *ln;
+
+    /* A worker can execute jobs of different types, so we must examine its jobs
+     * list and only count those we're interested in. */
+    pthread_mutex_lock(&bio_mutex[worker]);
+    listRewind(bio_jobs[worker],&li);
+    while((ln = listNext(&li))) {
+        bio_job *job = (bio_job *) listNodeValue(ln);
+        if (job->header.type == type)
+            val++;
+    }
```

Comment:
this loop can be very harmful in some cases.
maybe we only need to how many jobs each worker has, and it's not about the their type?
i.e. in many cases all we care about is if the job queue is empty or not, and that loop can be very long (e.g. for lazy free)

Reply:
You have a point. A good optimization would be simply return the `listLength` for workers that only handle a single job type. I'd like to avoid hard-coding that information in random places though, so the best way I can think of doing that is a loop that checks if the worker appears more than once in `bio_job_to_worker`. It's a bit of extra runtime code (assuming the compiler can't optimize that away), but either way the performance overhead should be pretty tiny. WDYT?

(Alternatively there might be some compile-time way to do the same computation that I'm missing?)

Reply:
well, i guess what you proposed would work, but i'm a bit scared that it'll someday stop working (or start working inefficiently) when more changes are made.

the fact is that currently anyone who's using it is using it as a boolean thing (to know if the queue is empty or non-empty).
so as soon as we found one, we can break out, no need for the count of listLength.

the other fact is that as far as i can tell, despite each (or actually just one) worker having multiple job types, we only care about the the worker being empty or non-empty.
so IMHO we can make a map of type to worker, and just find the job count (or boolean thing) of the corresponding worker.

maybe it'll require renaming the API to be on the safe side:
1. not being confusing
2. since we change the output (to boolean), we want it to fail if anyone used it in some redis fork